src/karts/controller/skidding_ai.cpp

```diff
@@ -1172,80 +1182,228 @@ void SkiddingAI::handleItems(const float dt)
-            // opportunity, since this kart won't envounter them anymore.
-            LinearWorld *lin_world = dynamic_cast<LinearWorld*>(World::getWorld());
-            if(m_time_since_last_shot > 3.0f &&
-                lin_world &&
-                lin_world->getKartLaps(m_kart->getWorldKartId())
-                                   == race_manager->getNumLaps()-1)
-            {
-                m_controls->setFire(true);
-                m_controls->setLookBack(true);
-                break;
-            }
-            break;   // POWERUP_BUBBLEGUM
-        }
+    int node = m_track_node;
+    float distance = 0;
+    std::vector<const Item *> items_to_collect;
+    std::vector<const Item *> items_to_avoid;
+
+    // 1) Filter and sort all items close by
+    // -------------------------------------
+    const float max_item_lookahead_distance = 10.f;
+    while(distance < max_item_lookahead_distance)
+    {
+        int n_index= DriveGraph::get()->getNode(node)->getIndex();
+        const std::vector<Item *> &items_ahead =
+            ItemManager::get()->getItemsInQuads(n_index);
+        for(unsigned int i=0; i<items_ahead.size(); i++)
+        {
+            evaluateItems(items_ahead[i],  kart_aim_direction,
+                          &items_to_avoid, &items_to_collect);
+        }   // for i<items_ahead;
+        distance += DriveGraph::get()->getDistanceToNext(node,
+                                                      m_successor_index[node]);
+        node = m_next_node_index[node];
+        // Stop when we have reached the last quad
+        if(node==last_node) break;
+    }   // while (distance < max_item_lookahead_distance)
+
+    //items_to_avoid and items_to_collect now contain the closest item information needed after
+    //What matters is (a) if the lists are void ; (b) if they are not, what kind of item it is
+   
+    switch( m_kart->getPowerup()->getType() )
+    {
+    // Level 2 : Use the shield immediately after a wait time
+    // Level 3 : Use the shield against flyable except cakes. Use the shield against bad attachments.
+    //           Use the bubble gum against an enemy close behind, except if holding a swatter.
+    // Level 4 : Level 3 and protect against cakes too
+    // Level 5 : Level 4 and let plunger hit (can use the shield after)
+    case PowerupManager::POWERUP_BUBBLEGUM:
+        {
+           Attachment::AttachmentType type = m_kart->getAttachment()->getType();
+           
+           if((ai_skill == 2) && (m_time_since_last_shot > 2.0f))
+           {
+               m_controls->setFire(true);
+               m_controls->setLookBack(false);
+               break;
+           }
+           
+           // Check if a flyable (cake, ...) is close. If so, use bubblegum
+           // as shield
+           if(ai_skill == 3) //don't protect against cakes
+           {
+              if( !m_kart->isShielded() && projectile_is_close
+                 && (projectile_types % 1000)/100 == 0)
```

Comment:
Instead of all these weird decimal additions, wouldn't a bitmask be simpler??

Reply:
I wanted to pass only one int arguments. The decimal trick was meant to allow this. I agree it's rather rough.

A bitmask would certainly work ; I simply have no experience using it.

Reply:
I looked up bitmasks.

I've not found clear examples and I think highly likely I would introduce some defect if I tried to do it.

However, if it is done, and some of my modulo/division tests are replaced by bitmasks, I think I should be able to imitate them for other cases.